src/shared/hostname-setup.c

```diff
@@ -114,17 +116,23 @@ int read_etc_hostname_stream(FILE *f, char **ret) {
                 if (IN_SET(line[0], '\0', '#'))
                         continue;
 
+                if (substitute_wildcards) {
+                        r = hostname_substitute_wildcards(line);
+                        if (r < 0)
+                                return r;
+                }
+
                 hostname_cleanup(line); /* normalize the hostname */
 
-                if (!hostname_is_valid(line, VALID_HOSTNAME_TRAILING_DOT)) /* check that the hostname we return is valid */
+                if (!hostname_is_valid(line, VALID_HOSTNAME_TRAILING_DOT|(substitute_wildcards ? 0 : VALID_HOSTNAME_QUESTION_MARK))) /* check that the hostname we return is valid */
```

Comment:
The line is slightly long. Please move the comment before the `if` condition or so.